stellar/txobserver/main.go

```diff
@@ -48,111 +59,148 @@ LedgerLoop:
+			ForLedger:     uint(o.ledgerSequence),
+			Cursor:        cursor,
+			Limit:         200,
+			IncludeFailed: true,
+		})
+		if err != nil {
+			return errors.Wrap(err, "error getting transactions")
 		}
 
-		// Process transactions
-		cursor := ""
-		for {
-			txs, err := o.client.Transactions(horizonclient.TransactionRequest{
-				ForLedger:     uint(o.ledgerSequence),
-				Cursor:        cursor,
-				Limit:         200,
-				IncludeFailed: true,
-			})
-			if err != nil {
-				o.log.WithFields(slog.F{
-					"error":  err,
-					"ledger": o.ledgerSequence,
-					"cursor": cursor,
-				}).Error("Error getting transactions")
-				time.Sleep(time.Second)
-				continue LedgerLoop
-			}
+		if len(txs.Embedded.Records) == 0 {
+			break
+		}
 
-			if len(txs.Embedded.Records) == 0 {
-				break
+		for _, tx := range txs.Embedded.Records {
+			otx, hashExists := outgoingBridgeTransactionsHashMap[tx.Hash]
+			if hashExists {
+				if tx.Successful {
+					otx.State = store.SuccessState
+					// TODO [important] update IncomingEthereumTransaction state
+					// as withdrawn so it's not possible to cancel it.
+				} else {
+					otx.State = store.FailedState
+				}
+				err := o.store.UpsertOutgoingStellarTransaction(context.TODO(), otx)
+				if err != nil {
+					return errors.Wrapf(err, "error upserting outgoing transaction: %s", otx.Hash)
+				}
 			}
 
-			for _, tx := range txs.Embedded.Records {
-				otx, hashExists := outgoingBridgeTransactionsMap[tx.Hash]
+			if tx.InnerTransaction != nil {
+				otx, hashExists := outgoingBridgeTransactionsHashMap[tx.InnerTransaction.Hash]
 				if hashExists {
 					if tx.Successful {
 						otx.State = store.SuccessState
+						// TODO [important] update IncomingEthereumTransaction state
+						// as withdrawn so it's not possible to cancel it.
 					} else {
 						otx.State = store.FailedState
 					}
 					err := o.store.UpsertOutgoingStellarTransaction(context.TODO(), otx)
 					if err != nil {
-						o.log.WithFields(slog.F{
-							"error": err,
-							"hash":  otx.Hash,
-						}).Error("Error upserting outgoing transactions")
-						time.Sleep(time.Second)
-						continue LedgerLoop
+						return errors.Wrapf(err, "error upserting outgoing transaction: %s", otx.Hash)
 					}
 				}
+			}
 
-				if tx.InnerTransaction != nil {
-					otx, hashExists := outgoingBridgeTransactionsMap[tx.InnerTransaction.Hash]
-					if hashExists {
-						if tx.Successful {
-							otx.State = store.SuccessState
-						} else {
-							otx.State = store.FailedState
-						}
+			seqnums, sourceExists := outgoingBridgeTransactionsSourceMap[tx.Account]
+			if sourceExists {
+				for seqnum, otx := range seqnums {
+					if tx.AccountSequence >= seqnum && otx.Hash != tx.Hash &&
+						(tx.InnerTransaction == nil ||
+							(tx.InnerTransaction != nil && otx.Hash != tx.InnerTransaction.Hash)) {
+						// Account sequence number was used for another transaction or bump thus otx transaction
```

Comment:
Are bridge nodes going to store the sequence numbers of all accounts on Stellar? If not it seems we might miss the fact that a transaction is irrevocably invalid, when it's so because of its sequence number, and the user will have to wait for the timeout in order to cancel and get a refund.

Reply:
Currently Starbridge only observes seqnums of accounts which have previously requested a corresponding inverse txs. The only case in which an account would need to wait instead of invalidating by seqnum is the seqnum provided to Starbridge was <= current seqnum and there were no txs happening in this account until the tx expiration time. Given that expiration time will be relatively small and the storage requirements for holding all accounts seqnums is quite large I think it's fine for now.

Reply:
Okay that makes sense.